src/Authentication.Abstractions/AzureEnvironment.cs

```diff
@@ -131,18 +153,185 @@ public class AzureEnvironment : IAzureEnvironment
                 BatchEndpointResourceId = AzureEnvironmentConstants.GermanBatchEndpointResourceId,
                 AdTenant = "Common"
             };
-            azureGermany.SetProperty(ExtendedEndpoint.AnalysisServicesEndpointSuffix, AzureEnvironmentConstants.GermanAnalysisServicesEndpointSuffix);
-            azureGermany.SetProperty(ExtendedEndpoint.AnalysisServicesEndpointResourceId, AzureEnvironmentConstants.GermanAnalysisServicesEndpointResourceId);
-            var result = new ConcurrentDictionary<string, AzureEnvironment>(StringComparer.InvariantCultureIgnoreCase);
 
+            var result = new ConcurrentDictionary<string, AzureEnvironment>(StringComparer.InvariantCultureIgnoreCase);
             result[EnvironmentName.AzureCloud] = azureCloud;
             result[EnvironmentName.AzureChinaCloud] = azureChina;
             result[EnvironmentName.AzureUSGovernment] = azureUSGovernment;
             result[EnvironmentName.AzureGermanCloud] = azureGermany;
+            SetExtendedProperties(result);
+            return result;
+        }
+
+        /// <summary>
+        /// Initializes cloud metadata dynamically from ARM.
+        /// </summary>
+        private static async Task<IDictionary<string, AzureEnvironment>> InitializeEnvironmentsFromArm(string armMetadataRequestUri)
+        {
+            HttpResponseMessage armResponseMessage = null;
+            await Policy
+                .Handle<HttpRequestException>()
+                .WaitAndRetryAsync(
+                    GetArmMetadataRetryCount,
+                    r => TimeSpan.FromSeconds(2 * r),
+                    (ex, ts, r) =>
+                        Debug.WriteLine(
+                            $"Could not retrieve ARM metadata. Attempt #{r}/{GetArmMetadataRetryCount}. Will try again in {ts.TotalSeconds} seconds. Exception={ex}"))
+                .ExecuteAsync(async () =>
+                {
+                    armResponseMessage = await Client.GetAsync(armMetadataRequestUri);
+                });
+
+            if (armResponseMessage == null || armResponseMessage?.StatusCode != HttpStatusCode.OK)
+            {
+                throw new Exception("Could not retrieve ARM metadata, compute management client cannot be initialized.");
```

Comment:
I think the error is not "compute management client cannot be initialized", more like "response is invalid"

Reply:
Updated error, thanks.